File: 0692-top-k-frequent-words/0692-top-k-frequent-words.java
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        Map<String, Integer> freq = new HashMap<>();

        for (String word : words) {
            freq.put(word, freq.getOrDefault(word, 0) + 1);
        }

        PriorityQueue<String> minHeap = new PriorityQueue<>((a, b) -> {
            int freqA = freq.get(a);
            int freqB = freq.get(b);

            if (freqA != freqB) {
                return freqA - freqB;
            }

            return b.compareTo(a);
        });

        for (String word : freq.keySet()) {
            minHeap.add(word);

            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }

        LinkedList<String> result = new LinkedList<>();

        while (!minHeap.isEmpty()) {
            result.addFirst(minHeap.poll());
        }

        return result;
    }
}

File: 0692-top-k-frequent-words/0692-top-k-frequent-words.kt
class Solution {
    fun topKFrequent(words: Array<String>, k: Int): List<String> {
        val freq = mutableMapOf<String, Int>()
        for (word in words) {
            freq[word] = freq.getOrDefault(word, 0) + 1
        }

        val minHeap = PriorityQueue<String> { a, b ->
            val freqDiff = freq[a]!! - freq[b]!!
            if (freqDiff != 0) freqDiff else b.compareTo(a)
        }

        for (word in freq.keys) {
            minHeap.add(word)
            if (minHeap.size > k) {
                minHeap.poll()
            }
        }

        val result = LinkedList<String>()
        while (minHeap.isNotEmpty()) {
            result.addFirst(minHeap.poll())
        }

        return result
    }
}